File: java/src/edu/berkeley/cs/util/TreeMap.java
package edu.berkeley.cs.util;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class TreeMap<Key extends Comparable<Key>, Value> implements Map<Key, Value>, Iterable<Key> {
  protected Node root;
  private int size;

  protected class Node {
    public Key key;
    public Value value;
    public Node left;
    public Node right;

    public Node(Key key, Value value) {
      this.key = key;
      this.value = value;
    }
  }

  /** Reset this container as if it was newly created with no elements */
  @Override
  public void clear() {
    root = null;
    size = 0;
  }

  /** @return true if this container contains no elements */
  @Override
  public boolean isEmpty() {
    return size() == 0;
  }

  /** @return the number of elements in this container */
  @Override
  public int size() {
    return size;
  }

  /**
   * @param key The element whose presence in this container is to be tested. If key is null, throw
   *     an IllegalArgumentException.
   * @return true if this container contains the specified element
   */
  @Override
  public boolean contains(Key key) {
    if (key == null) {
      throw new IllegalArgumentException("argument to contains() is null");
    }

    return get(key) != null;
  }

  /**
   * @param key The element whose presence in this container is to be tested. If key is null, throw
   *     an IllegalArgumentException.
   * @return the value associated with the input key, if it exists in the container (null otherwise)
   */
  @Override
  public Value get(Key key) {
    // TODO: complete this function
    return null;
  }

  /**
   * Associate a key with a value. If the key already exists in the container, this method
   * overwrites the previous value with the new one.
   *
   * @param key the key to be inserted into the container (cannot be null). If a null key is
   *     provided, throw an IllegalArgumentException.
   * @param value the value associated with the input key (deletes key if null)
   */
  @Override
  public void put(Key key, Value value) {
    if (key == null) {
      throw new IllegalArgumentException("calls put() with a null key");
    }

    if (value == null) {
      delete(key);
      return;
    }

    size++;
    // TODO: complete this function
  }

  /**
   * Delete the minimum element in the container. If is called on an empty container, throw a
   * NoSuchElementException.
   */
  public void deleteMin() {
    if (isEmpty()) {
      throw new NoSuchElementException("Symbol table underflow");
    }

    root = deleteMin(root);
    size--;
  }

  /**
   * @param node the root of the subtree to delete the minimum from
   * @return the new left node for this subtree root
   */
  private Node deleteMin(Node node) {
    // TODO: complete this function
    return null;
  }

  /**
   * Delete the maximum element in the container. If called on an empty container, throw a
   * NoSuchElementException.
   */
  public void deleteMax() {
    if (isEmpty()) {
      throw new NoSuchElementException("Symbol table underflow");
    }

    root = deleteMax(root);
    size--;
  }

  /**
   * @param node the root of the subtree to delete the maximum from
   * @return the new right node for this subtree root
   */
  private Node deleteMax(Node node) {
    // TODO: complete this function
    return null;
  }

  /**
   * @param key the key to delete from the container (the associated value is also deleted). If key
   *     is null, throw an IllegalArgumentException.
   */
  @Override
  public void delete(Key key) {
    if (key == null) {
      throw new IllegalArgumentException("calls delete() with a null key");
    }

    root = delete(root, key);
    size--;
  }

  /**
   * @param node the subtree root to delete from
   * @param key the key to delete from the subtree (the associated value is also deleted)
   */
  private Node delete(Node node, Key key) {
    if (node == null) {
      return null;
    }

    int cmp = key.compareTo(node.key);
    if (cmp < 0) {
      node.left = delete(node.left, key);
    } else if (cmp > 0) {
      node.right = delete(node.right, key);
    } else {
      if (node.right == null) {
        return node.left;
      }

      if (node.left == null) {
        return node.right;
      }

      Node t = node;
      node = min(t.right);
      node.right = deleteMin(t.right);
      node.left = t.left;
    }

    return node;
  }

  /**
   * @return the minimum value key in the container. If called on an empty container, throw a
   *     NoSuchElementException.
   */
  public Key min() {
    if (isEmpty()) {
      throw new NoSuchElementException("calls min() with empty symbol table");
    }

    return min(root).key;
  }

  /**
   * @param node the root of the subtree to find the minimum from
   * @return the minimum node within this subtree
   */
  private Node min(Node node) {
    // TODO: complete this function
    return null;
  }

  /**
   * @return the maximum value key in the container. If called on an empty container, throw a
   *     NoSuchElementException.
   */
  public Key max() {
    if (isEmpty()) {
      throw new NoSuchElementException("calls max() with empty symbol table");
    }

    return max(root).key;
  }

  /**
   * @param node the root of the subtree to find the maximum from
   * @return the maximum node within this subtree
   */
  private Node max(Node node) {
    // TODO: complete this function
    return null;
  }

  /** @return true if the container exhibits the binary search tree property at all nodes */
  public boolean isBST() {
    return isBST(root, null, null);
  }

  /**
   * @return true if the container exhibits the binary search tree property at the subtree rooted at
   *     node.
   */
  private boolean isBST(Node node, Key min, Key max) {
    if (node == null) {
      return true;
    }

    if (min != null && node.key.compareTo(min) <= 0) {
      return false;
    }

    if (max != null && node.key.compareTo(max) >= 0) {
      return false;
    }

    return isBST(node.left, min, node.key) && isBST(node.right, node.key, max);
  }

  @Override
  public Iterator<Key> iterator() {
    Queue<Key> keys = new Queue<>();
    Queue<Node> queue = new Queue<>();
    queue.enqueue(root);

    while (!queue.isEmpty()) {
      Node node = queue.dequeue();
      if (node == null) {
        continue;
      }

      keys.enqueue(node.key);
      queue.enqueue(node.left);
      queue.enqueue(node.right);
    }

    return new Iterator<Key>() {
      private Iterator<Key> itr = keys.iterator();

      @Override
      public boolean hasNext() {
        return itr.hasNext();
      }

      @Override
      public Key next() {
        return itr.next();
      }
    };
  }
}

File: python/src/util/treemap.py
from python.src.util.map import Map
from python.src.util.queue import Queue


class TreeMap(Map):

    class Node:

        def __init__(self, key, value):
            self.key = key
            self.value = value
            self.left = None
            self.right = None

    def __init__(self):
        self.root = None
        self.num_elements = 0

    def clear(self):
        self.root = None
        self.num_elements = 0

    def is_empty(self):
        return self.size() == 0

    def size(self):
        return self.num_elements

    def contains(self, key):
        if key is None:
            raise Exception("argument to contains() is None")

        return self.get(key) is not None

    def get(self, key):
        return self.__get(self.root, key)

    def __get(self, node, key):
        # TODO: complete this function
        return None

    def put(self, key, value):
        if key is None:
            raise Exception("calls put() with a None key")

        if value is None:
            self.delete(key)
            return

        self.root = self.__put(self.root, key, value)
        self.num_elements += 1

    def __put(self, node, key, value):
        # TODO: complete this function
        return None

    def delete_min(self):
        if self.is_empty():
            raise Exception("Symbol table underflow")

        self.root = self.__delete_min(self.root)
        self.num_elements -= 1

    def __delete_min(self, node):
        # TODO: complete this function
        return None

    def delete_max(self):
        if self.is_empty():
            raise Exception("Symbol table underflow")

        self.root = self.__delete_max(self.root)
        self.num_elements -= 1

    def __delete_max(self, node):
        # TODO: complete this function
        return None

    def delete(self, key):
        if key is None:
            raise Exception("calls delete() with a None key")

        self.root = self.__delete(self.root, key)
        self.num_elements -= 1

    def __delete(self, node, key):
        if node is None:
            return None

        if key < node.key:
            node.left = self.__delete(node.left, key)
        elif key > node.key:
            node.right = self.__delete(node.right, key)
        else:
            if node.right is None:
                return node.left

            if node.left is None:
                return node.right

            t = node
            node = self.__min(t.right)
            node.right = self.__delete_min(t.right)
            node.left = t.left

        return node

    def min(self):
        if self.is_empty():
            raise Exception("calls min() with empty symbol table")

        return self.__min(self.root).key

    def __min(self, node):
        # TODO: complete this function
        return None

    def max(self):
        if self.is_empty():
            raise Exception("calls max() with empty symbol table")

        return self.__max(self.root).key

    def __max(self, node):
        # TODO: complete this function
        return None

    def is_bst(self):
        return self.__is_bst(self.root, None, None)

    def __is_bst(self, node, minimum, maximum):
        if node is None:
            return True

        if minimum is not None and node.key <= minimum:
            return False

        if maximum is not None and node.key >= maximum:
            return False

        return self.__is_bst(node.left, minimum, node.key) and self.__is_bst(
            node.right, node.key, maximum)

    class TreeMapIterator:

        def __init__(self, root):
            keys = Queue()
            queue = Queue()
            queue.enqueue(root)

            while not queue.is_empty():
                node = queue.dequeue()
                if node is None:
                    continue

                queue.enqueue(node.left)
                keys.enqueue(node.key)
                queue.enqueue(node.right)

            self.iterator = iter(keys)

        def __iter__(self):
            return self.iterator

        def __next__(self):
            return next(self.iterator)

    def __iter__(self):
        return self.TreeMapIterator(self.root)
